File: java/minimumDominoRotationsForEqualRow.java
class Solution {
    public int minDominoRotations(int[] A, int[] B) {
        //keep track of the eligible numbers (that is, the numbers that appear in every such domino)
        Set<Integer> eligible = new HashSet<>();
        eligible.add(A[0]);
        eligible.add(B[0]);
        //for each eligible number, keep track of the number of times it appears in the upper and lower positions
        Map<Integer, Integer[]> indexCount = new HashMap<>();
        for (int el : eligible) {
            indexCount.put(el, new Integer[]{ 0, 0 });
        }
        //iterate through the dominos
        for (int i = 0; i < A.length; i++) {
            //remove from the eligible set any numbers that don't appear
            Set<Integer> toRemove = new HashSet<>();
            for (int el : eligible) {
                if (el != A[i] && el != B[i]) {
                    toRemove.add(el);
                }
            }
            for (int el : toRemove) {
                eligible.remove(el);
            }
            //if there are no more eligible numbers, there is no solution
            if (eligible.size() == 0) {
                return -1;
            }
            //increment upper and lower indices for each eligible number
            if (eligible.contains(A[i])) {
                Integer[] curIndexCount = indexCount.get(A[i]);
                curIndexCount[1]++;
            }
            if (eligible.contains(B[i])) {
                Integer[] curIndexCount = indexCount.get(B[i]);
                curIndexCount[0]++;
            }
        }
        //find the greatest number of indicies among eligible numbers and positions, and substract it 
        //from the number of dominos to get the minimum number of rotations
        int minRotations = A.length;
        for (int el : eligible) {
            for (int i = 0; i < 2; i++) {
                minRotations = Math.min(minRotations, A.length - indexCount.get(el)[i]);
            }
        }
        return minRotations;
    }
}

File: python/minimum_domino_rotations_for_equal_row.py
class Solution:
    def minDominoRotations(self, A: List[int], B: List[int]) -> int:
        #keep track of which numbers are still eligible: that is, the numbers that have appeared at least once on every domino thus far
        eligible = { A[0], B[0] }
        #for each eligible number, keep track of the number of indices it appear in the upper and lower levels
        class Solution:
        upper_and_lower_indices = dict()
        for num in eligible:
            upper_and_lower_indices[num] = { "upper": 0, "lower": 0 }
        #iterate through each domino
        for i in range(len(A)):
            to_remove = set()
            #if a number doesn't appear in the domino, it is no longer eligible
            for el in eligible:
                if el not in { A[i], B[i] }:
                    to_remove.add(el)
            for el in to_remove:
                eligible.remove(el)
            #if there are no eligible numbers, there is no solution
            if len(eligible) == 0:
                return -1
            #update indices count for remaining elibile numbers
            if A[i] in eligible:
                upper_and_lower_indices[A[i]]["upper"] += 1
            if B[i] in eligible:
                upper_and_lower_indices[B[i]]["lower"] += 1
        #determine the minimum number of rotations by finding the eligible number with the greatest number of indices in either position
        min_rotations = len(A)
        for el in eligible:
            for position in ["upper", "lower"]:
                min_rotations = min(min_rotations, len(A) - upper_and_lower_indices[el][position])
        return min_rotations